packages/sphinx-external-toc/src/toc.ts

```diff
@@ -0,0 +1,168 @@
+  tocUpdated[containsPart ? 'parts' : topItemsKey] = firstItems;
+
+  // Write root
+  const { file: root, ...toc } = tocUpdated;
+  if (root === undefined) {
+    throw new Error("no top-level 'file' key found");
+  }
+  toc['root'] = root;
+
+  // Ensure we don't mix top-level key types
+  const topLevelKeys = ['parts', 'chapters', 'sections'].filter((item) => item in toc);
+  if (topLevelKeys.length > 1) {
+    throw new Error(`There is more than one top-level key: ${topLevelKeys}`);
+  }
+
+  // Deduce the TOC format
+  if (topLevelKeys.length) {
+    const fileFormat = {
+      parts: 'jb-book',
+      chapters: 'jb-book',
+      sections: 'jb-article',
+    }[topLevelKeys[0]];
+    toc['format'] = fileFormat;
+  }
+
+  // Do we have a singular "default" subtree (indicated by "entries" key)
+  const hasDefaultSubtree = 'entries' in toc || 'sections' in toc || 'chapters' in toc;
+
+  // Lower numbering
+  const numbered = toc['numbered'];
+  delete toc['numbered'];
+  if (numbered !== undefined) {
+    // Default subtree
+    if (hasDefaultSubtree) {
+      toc['options'] = { numbered: numbered };
+    }
+    // Child subtrees
+    else {
+      const subtrees = (toc['subtrees'] ?? toc['parts'] ?? []) as Record<string, unknown>[];
+      for (const subtree of subtrees) {
+        subtree['numbered'] = numbered;
+      }
+    }
+  }
+
+  // Lower title
+  const title = toc['title'];
+  delete toc['title'];
+  if (title !== undefined) {
+    // Only set title for single default subtree
+    if (hasDefaultSubtree) {
+      const options = (toc['options'] ?? (toc['options'] = {})) as Record<string, unknown>;
+      options['caption'] = title;
+    }
+  }
+
+  // Rename "part" to "caption"
+  const adjustConfig = (obj: Record<string, unknown>) => {
+    if ('chapters' in obj && 'sections' in obj) {
+      throw new Error(`both 'chapters' and 'sections' in same table: ${obj}`);
+    }
+    const caption = obj['part'] ?? obj['chapter'];
+    delete obj['part'];
+    delete obj['chapter'];
+
+    if (caption !== undefined) {
+      obj['caption'] = caption;
+    }
+    for (const key of ['parts', 'chapters', 'sections', 'entries']) {
+      if (key in obj) {
+        const children = obj[key] as Record<string, unknown>[];
+        children.forEach(adjustConfig);
+      }
+    }
+  };
+
+  adjustConfig(toc);
+  return toc;
+}
+
+/**
+ * Parse a sphinx-external-toc table of contents
+ *
+ * @param contents: raw TOC yaml
+ */
+export function parseTOC(contents: string): { toc: TOC; didUpgrade: boolean } {
+  let toc: any;
+  try {
+    toc = yaml.load(contents) as any;
+  } catch (err) {
+    throw new Error(`Unable to parse TOC yaml`);
+  }
+  let didUpgrade = false;
+
+  if (Array.isArray(toc)) {
+    toc = upgradeOldJupyterBookTOC(toc);
+    console.log('upgraded', JSON.stringify(toc));
+    didUpgrade = true;
+  }
+
+  const ajv = new Ajv.default();
+  const validate = ajv.compile(schema);
+  if (!validate(toc)) {
+    throw new Error(
```

Comment:
Worth adding `validate.errors` to this message? I think you were saying those errors were sometimes totally undecipherable because of all the complexity with unions, so probably not, I suppose.

Reply:
I'm not sure -- as you note, I found that the errors are fairly non-intuitive. I think for now it might be sufficient just to complain about an invalid TOC.